internal/worker/sshsession/worker.go

```diff
@@ -0,0 +1,326 @@
+		return errors.NotValidf("nil ConnectionGetter")
+	}
+	if cfg.EphemeralKeysUpdater == nil {
+		return errors.NotValidf("nil EphemeralKeysUpdater")
+	}
+
+	return nil
+}
+
+// sshSessionWorker is a worker that enables SSH connections to a machine.
+type sshSessionWorker struct {
+	catacomb catacomb.Catacomb
+
+	logger               Logger
+	machineId            string
+	facadeClient         FacadeClient
+	connectionGetter     ConnectionGetter
+	ephemeralKeysUpdater authenticationworker.EphemeralKeysUpdater
+}
+
+// NewWorker returns an SSH session worker.
+func NewWorker(cfg WorkerConfig) (worker.Worker, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &sshSessionWorker{
+		logger:               cfg.Logger,
+		machineId:            cfg.MachineId,
+		facadeClient:         cfg.FacadeClient,
+		connectionGetter:     cfg.ConnectionGetter,
+		ephemeralKeysUpdater: cfg.EphemeralKeysUpdater,
+	}
+
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	})
+
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// loop starts the workers main loop.
+func (w *sshSessionWorker) loop() error {
+	connRequestWatcher, err := w.facadeClient.WatchSSHConnRequest(w.machineId)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := w.catacomb.Add(connRequestWatcher); err != nil {
+		return errors.Trace(err)
+	}
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case changes := <-connRequestWatcher.Changes():
+			for _, connId := range changes {
+				go func() {
+					requestContext := w.catacomb.Context(context.Background())
+					err := w.handleConnection(requestContext, connId)
+					if err != nil {
+						w.logger.Errorf("Failed to handle connection %q: %v", connId, err)
+					}
+				}()
+			}
+		}
+	}
+}
+
+// Kill implements the worker.Worker interface.
+func (w *sshSessionWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait implements the worker.Worker interface.
+func (w *sshSessionWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+// handleConnection handles initiating a reverse SSH connection to the controller
+// and piping it to the local sshd of the machine agent's machine.
+//
+// This function does the following:
+//  1. Gets the controllers address and ephemeral public key for the connection.
+//  2. Verifies the address is known to this machine agent.
+//  3. Adds the ephemeral public key.
+//  4. Dials the controllers SSH server expecting an SSH connection to come
+//     back from the controller's SSH server.
+//  5. Pipes the connection to the local sshd.
+//  6. On connection close, removes the ephemeral public key.
+func (w *sshSessionWorker) handleConnection(ctx context.Context, connID string) error {
+	reqParams, err := w.facadeClient.GetSSHConnRequest(connID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	ctrlAddress := reqParams.ControllerAddresses.Values()[0]
```

Comment:
I suggest we loop over these values, trying each one. But let's address in a follow-up.